hgshah/cmput404-project
# Follow a remote author

Closes https://trello.com/c/KU4SnMEt/125-as-an-author-i-want-to-follow-remote-authors-or-authors-that-are-hosted-or-owned-by-remote-nodes

## Changes
- Prefilling database to make it easier every time we wipe our database during drastic Model changes (like right now LOL)
- Add remote node credentials via environment variable (todo: link to doc)
- For local authors, host is now useless since it's now based on the base.CURRENT_DOMAIN environment variable
- Make a follow request for remote author

## TODO
- [x] Make a Follow request on our end too uhhh ☠
  - local_follow_url
  - remote_follow_url
- [x] Document post remote followers
- [x] Document config vars
- [x] Document NodeConfigBase

## Testing
Let's say we want our super user in potato-oomfie-to follow the target user in turnip-oomfie-1
1. Login as **super** in https://turnip-oomfie-1.herokuapp.com/admin/ (password is the same as the username...)
2. Go to Follows and delete the Follow object where **super** (from potato-oomfie) wants to follow **target** (from turnip-oomfie-1), if it exists
3. Login as super in https://potato-oomfie.herokuapp.com/admin/
4. To see the followers of target via potato-oomfie's endpoint, go to https://potato-oomfie.herokuapp.com/authors/ef9e65e4-cbf8-457e-9ec8-aea8599dbbb4/followers/?node-target=turnip-oomfie-1.herokuapp.com.
     - This follows the pattern: `{local-service}/authors/{remote-author-id}/followers/?node-target={remote-node}`
     - This follows the pattern: `{super's service}/authors/{target's id}/followers/?node-target={the domain of the nodewhere target belongs}`
5. To make a follow request, we make a POST request at that same endpoint: https://potato-oomfie.herokuapp.com/authors/ef9e65e4-cbf8-457e-9ec8-aea8599dbbb4/followers/?node-target=turnip-oomfie-1.herokuapp.com

It should return something like this:
```json
{
    "type": "Follow",
    "id": 4,
    "summary": "super wants to follow {'type': 'author', 'id': 'http://turnip-oomfie-1.herokuapp.com/authors/606f18b1-d8cd-4e26-bf39-6a7dc2f16cf8', 'url': 'http://turnip-oomfie-1.herokuapp.com/authors/606f18b1-d8cd-4e26-bf39-6a7dc2f16cf8', 'host': '[turnip-oomfie-1.herokuapp.com](http://turnip-oomfie-1.herokuapp.com/)', 'displayName': 'target', 'github': 'https://github.com/target/', 'profileImage': ''}",
    "hasAccepted": false,
    "object": {
        "type": "author",
        "id": "http://turnip-oomfie-1.herokuapp.com/authors/606f18b1-d8cd-4e26-bf39-6a7dc2f16cf8",
        "url": "http://turnip-oomfie-1.herokuapp.com/authors/606f18b1-d8cd-4e26-bf39-6a7dc2f16cf8",
        "host": "[turnip-oomfie-1.herokuapp.com](http://turnip-oomfie-1.herokuapp.com/)",
        "displayName": "target",
        "github": "https://github.com/target/",
        "profileImage": ""
    },
    "actor": {
        "type": "author",
        "id": "http://potato-oomfie.herokuapp.com/authors/7a50a4b3-7a0b-4891-ab3e-da51c9119443",
        "url": "http://potato-oomfie.herokuapp.com/authors/7a50a4b3-7a0b-4891-ab3e-da51c9119443",
        "host": "[potato-oomfie.herokuapp.com](http://potato-oomfie.herokuapp.com/)",
        "displayName": "super",
        "github": "https://github.com/super/",
        "profileImage": ""
    },
    "localUrl": "http://potato-oomfie.herokuapp.com/follows/4",
    "remoteUrl": "http://turnip-oomfie-1.herokuapp.com/follows/23"
}
```

## Remote Follow Implementation

1. super makes an authenticated POST request to the server it belongs to, potato-oomfie.herokuapp.com
2. The server sees that this call was made by a user and that there's a node-target query param indicating it will make a follow request to a remote server
3. The server looks at the implementations based on the domain, and will try to use the logic in it
    - To be more precise: the server will use the key `potato-oomfie.herokuapp.com` from the node-target query param, then will use the config returned, and call the `post_local_follow_remote` function in that config object
4. The server makes a BasicAuth POST call to `turnip-oomfie-1`, with a payload `actor=super's author url` at the endpoint [https://turnip-oomfie-1.herokuapp.com/authors/606f18b1-d8cd-4e26-bf39-6a7dc2f16cf8/followers/](https://turnip-oomfie-1.herokuapp.com/authors/606f18b1-d8cd-4e26-bf39-6a7dc2f16cf8/followers/)
    - Notice that the two differences here is the domain and that there's no query param node-target appended to it
5. The `turnip-oomfie-1` server will see that this call was made by a **node**, so it will do some slight adjustment to the normal make a follow request logic
    - Compare this with step 2
6. `turnip-oomfie-1` will make a Follow object based on the actor payload for super (in potato) and the author id for target (in turnip)
7. `turnip-oomfie-1` will send back a Follow json and 200 to `potato-oomfie`
8. `potato-oomfie` will save a copy of this Follow object. In addition, it will save the url where the Follow resource exists in `turnip-oomfie-1` indicating that the main source of truth is in `turnip-oomfie-1`
9. `potato-oomfie` will send back that same Folllow json and 200 to the super's client

- Step 1 to 8 is how a local author would follow a remote author
- Step 4 to 7 is how a remote server would communicate with our local server

## Contrasting with a local author follow a local author

Let's say super wants to follow banana

1. super's client will call the server with an authenticated `POST service/authors/{banana_id}/followers`
2. the server will see that a user (super) made this call and that there was no node_query param
3. the server will make a Follow object with **super** as the **actor** and **banana** as the **target**
4. the server will send back a Follow json and 200 to super's client

- Compare the step 1 and 2 here, with the step 1 and 2 in the remote flow
- Compare the step 3 and 4 here, with the step 6 and 7 in the remote flow
- Step 8 in the local-remote flow is a special version of step 3 in the local-local flow

**Both of these calls will return the same format:** see [the drf-spectacular documentation](http://potato-oomfie.herokuapp.com/api/schema/redoc/#tag/follows/operation/authors_followers_create)
